File: models/spare.py
# -*- coding: utf-8 -*-
"""备件 / 库存 / 采购 / 销售单模型"""
from datetime import datetime
from models.base import db


# ============================
# 备件管理（四合一）
# ============================

class SparePart(db.Model):
    """备件档案"""
    __tablename__ = 'spare_parts'
    id = db.Column(db.Integer, primary_key=True)
    code = db.Column(db.String(64), unique=True, default='')
    name = db.Column(db.String(128), nullable=False)
    category = db.Column(db.String(64), default='')
    specification = db.Column(db.String(128), default='')
    unit = db.Column(db.String(16), default='个')
    min_stock = db.Column(db.Integer, default=0)
    remark = db.Column(db.Text, default='')
    # V6 新增业务字段
    brand = db.Column(db.String(64), default='')             # 品牌
    model = db.Column(db.String(64), default='')             # 型号
    parameters = db.Column(db.Text, default='')              # 详细参数
    manufacturer = db.Column(db.String(64), default='')      # 厂家
    image_path = db.Column(db.String(512), default='')       # 备件图片
    serial_number = db.Column(db.String(128), default='')    # 序列号
    reference_price = db.Column(db.Float, default=0.0)       # 采购参考价
    warranty_months = db.Column(db.Integer, default=0)       # 保修期（月）
    created_at = db.Column(db.DateTime, default=datetime.utcnow)


class SpareStock(db.Model):
    """备件库存"""
    __tablename__ = 'spare_stocks'
    id = db.Column(db.Integer, primary_key=True)
    spare_part_id = db.Column(db.Integer, db.ForeignKey('spare_parts.id'), nullable=False, index=True)
    location = db.Column(db.String(128), default='')
    quantity = db.Column(db.Integer, default=0)
    unit_price = db.Column(db.Float, default=0.0)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    spare_part_rel = db.relationship('SparePart', backref='stocks')


class PurchaseOrder(db.Model):
    """采购入库单"""
    __tablename__ = 'purchase_orders'
    id = db.Column(db.Integer, primary_key=True)
    number = db.Column(db.String(64), default='')
    spare_part_id = db.Column(db.Integer, db.ForeignKey('spare_parts.id'), nullable=True)
    supplier_name = db.Column(db.String(128), default='')
    quantity = db.Column(db.Integer, default=0)
    unit_price = db.Column(db.Float, default=0.0)
    total = db.Column(db.Float, default=0.0)
    purchase_date = db.Column(db.Date, nullable=True)
    operator = db.Column(db.String(64), default='')
    remark = db.Column(db.Text, default='')
    created_at = db.Column(db.DateTime, default=datetime.utcnow)

    spare_part_rel = db.relationship('SparePart', backref='purchases')


class SalesOrder(db.Model):
    """销售出库单"""
    __tablename__ = 'sales_orders'
    id = db.Column(db.Integer, primary_key=True)
    number = db.Column(db.String(64), default='')
    spare_part_id = db.Column(db.Integer, db.ForeignKey('spare_parts.id'), nullable=True)
    customer_id = db.Column(db.Integer, db.ForeignKey('customers.id'), nullable=True)
    quantity = db.Column(db.Integer, default=0)
    unit_price = db.Column(db.Float, default=0.0)
    total = db.Column(db.Float, default=0.0)
    sales_date = db.Column(db.Date, nullable=True)
    operator = db.Column(db.String(64), default='')
    invoice_number = db.Column(db.String(128), default='')
    remark = db.Column(db.Text, default='')
    created_at = db.Column(db.DateTime, default=datetime.utcnow)

    spare_part_rel = db.relationship('SparePart', backref='sales')
    customer_rel = db.relationship('Customer', backref='sales_orders')


class StockMovement(db.Model):
    """库存流水（单一账本：采购/销售/冲销/盘点调整全部走此表，来源可追溯）"""
    __tablename__ = 'stock_movements'
    id = db.Column(db.Integer, primary_key=True)
    spare_part_id = db.Column(db.Integer, db.ForeignKey('spare_parts.id'), nullable=False, index=True)
    movement_type = db.Column(db.String(16), nullable=False)  # purchase/sales/purchase_cancel/sales_cancel/adjust
    quantity = db.Column(db.Integer, default=0)                # 正=入库 负=出库
    balance_after = db.Column(db.Integer, default=0)           # 变动后该备件总库存
    location = db.Column(db.String(128), default='')
    source_id = db.Column(db.Integer, default=None)            # 关联单据 id（采购/销售/库存行）
    operator = db.Column(db.String(64), default='')
    remark = db.Column(db.Text, default='')
    created_at = db.Column(db.DateTime, default=datetime.utcnow, index=True)

    spare_part_rel = db.relationship('SparePart', backref='movements')


class SpareBorrow(db.Model):
    """备件借用（借出登记 → 归还回库；借出扣库存、归还回补，流水走 stock_movements）"""
    __tablename__ = 'spare_borrows'
    id = db.Column(db.Integer, primary_key=True)
    spare_part_id = db.Column(db.Integer, db.ForeignKey('spare_parts.id'), nullable=False, index=True)
    borrower = db.Column(db.String(64), nullable=False)       # 借用人
    borrower_phone = db.Column(db.String(32), default='')
    quantity = db.Column(db.Integer, nullable=False)          # 借出数量（>0）
    location = db.Column(db.String(128), default='')          # 出借库位
    borrow_date = db.Column(db.Date, default=datetime.utcnow().date, index=True)
    expected_return_date = db.Column(db.Date, nullable=True)  # 预计归还日
    return_date = db.Column(db.Date, nullable=True)           # 实际归还日
    status = db.Column(db.String(16), default='借用中', index=True)  # 借用中/已归还/逾期
    operator = db.Column(db.String(64), default='')           # 经办人
    remark = db.Column(db.Text, default='')
    created_at = db.Column(db.DateTime, default=datetime.utcnow)

    spare_part_rel = db.relationship('SparePart', backref='borrows')

File: frontend/src/api/spare.ts
import request from '@/utils/request'
import type { PageResult } from '@/types'

export interface SparePart {
  id: number
  code: string
  name: string
  category: string
  brand: string
  model: string
  specification: string
  unit: string
  min_stock: number
  reference_price: number
  warranty_months: number
  manufacturer: string
  serial_number: string
  remark: string
  total_stock: number
  stock_alert: boolean
  stock_alert_label: string
  created_at: string
  stocks?: SpareStock[]
  purchases?: PurchaseOrderItem[]
  sales?: SalesOrderItem[]
}

export interface SpareStock {
  id: number
  spare_part_id: number
  spare_part_name: string
  location: string
  quantity: number
  unit_price: number
  updated_at: string
}

export interface PurchaseOrderItem {
  id: number
  number: string
  spare_part_id: number
  spare_part_name: string
  supplier_name: string
  quantity: number
  unit_price: number
  total: number
  purchase_date: string
  operator: string
  remark: string
  created_at: string
}

export interface SalesOrderItem {
  id: number
  number: string
  spare_part_id: number
  spare_part_name: string
  customer_id: number | null
  customer_name: string
  quantity: number
  unit_price: number
  total: number
  sales_date: string
  operator: string
  invoice_number: string
  remark: string
  created_at: string
}

export interface SpareDicts {
  spare_parts: { id: number; name: string }[]
  customers: { id: number; name: string }[]
  categories: string[]
}

export interface SparePartQuery {
  page?: number
  page_size?: number
  search?: string
  category?: string
}

export function fetchSpareParts(params: SparePartQuery) {
  return request<PageResult<SparePart>>({ url: '/api/spare-parts', method: 'GET', params })
}

export function fetchSparePart(id: number) {
  return request<SparePart>({ url: `/api/spare-parts/${id}`, method: 'GET' })
}

export function createSparePart(data: Record<string, unknown>) {
  return request<{ id: number }>({ url: '/api/spare-parts', method: 'POST', data })
}

export function updateSparePart(id: number, data: Record<string, unknown>) {
  return request<{ id: number }>({ url: `/api/spare-parts/${id}`, method: 'PUT', data })
}

export function deleteSparePart(id: number) {
  return request<null>({ url: `/api/spare-parts/${id}`, method: 'DELETE' })
}

export function fetchSpareStocks(params: Record<string, unknown>) {
  return request<PageResult<SpareStock>>({ url: '/api/spare-stocks', method: 'GET', params })
}

export function createSpareStock(data: Record<string, unknown>) {
  return request<{ id: number }>({ url: '/api/spare-stocks', method: 'POST', data })
}

export function updateSpareStock(id: number, data: Record<string, unknown>) {
  return request<{ id: number }>({ url: `/api/spare-stocks/${id}`, method: 'PUT', data })
}

export function deleteSpareStock(id: number) {
  return request<null>({ url: `/api/spare-stocks/${id}`, method: 'DELETE' })
}

export function fetchPurchaseOrders(params: Record<string, unknown>) {
  return request<PageResult<PurchaseOrderItem>>({ url: '/api/purchase-orders', method: 'GET', params })
}

export function createPurchaseOrder(data: Record<string, unknown>) {
  return request<{ id: number }>({ url: '/api/purchase-orders', method: 'POST', data })
}

export function deletePurchaseOrder(id: number) {
  return request<null>({ url: `/api/purchase-orders/${id}`, method: 'DELETE' })
}

export function fetchSalesOrders(params: Record<string, unknown>) {
  return request<PageResult<SalesOrderItem>>({ url: '/api/sales-orders', method: 'GET', params })
}

export function createSalesOrder(data: Record<string, unknown>) {
  return request<{ id: number }>({ url: '/api/sales-orders', method: 'POST', data })
}

export function deleteSalesOrder(id: number) {
  return request<null>({ url: `/api/sales-orders/${id}`, method: 'DELETE' })
}

export function fetchSpareDicts() {
  return request<SpareDicts>({ url: '/api/dicts/spare', method: 'GET' })
}
